crates/bevy_image/src/image.rs

```diff
@@ -1113,25 +1114,68 @@ impl Image {
     }
 
     /// Get a reference to the data bytes where a specific pixel's value is stored
-    #[inline(always)]
+    #[inline]
     pub fn pixel_bytes(&self, coords: UVec3) -> Option<&[u8]> {
         let len = self.texture_descriptor.format.pixel_size();
         self.pixel_data_offset(coords)
             .map(|start| &self.data[start..(start + len)])
     }
 
     /// Get a mutable reference to the data bytes where a specific pixel's value is stored
-    #[inline(always)]
+    #[inline]
     pub fn pixel_bytes_mut(&mut self, coords: UVec3) -> Option<&mut [u8]> {
         let len = self.texture_descriptor.format.pixel_size();
         self.pixel_data_offset(coords)
             .map(|start| &mut self.data[start..(start + len)])
     }
 
+    /// Get an iterator over the row bytes for a specific rect of pixels.
+    #[inline]
+    pub fn rect_bytes(&self, rect: URect, z: u32) -> Option<RectBytes<'_>> {
+        let (data_range, chunk_len, row_range) = self.rect_bytes_inner(rect, z)?;
+        Some(RectBytes {
+            data: self.data[data_range].chunks_exact(chunk_len),
+            range: row_range,
+            format: self.texture_descriptor.format,
+        })
+    }
+
+    /// Get an iterator over the row bytes for a specific rect of pixels.
+    #[inline]
+    pub fn rect_bytes_mut(&mut self, rect: URect, z: u32) -> Option<RectBytesMut<'_>> {
+        let (data_range, chunk_len, row_range) = self.rect_bytes_inner(rect, z)?;
+        Some(RectBytesMut {
+            data: self.data[data_range].chunks_exact_mut(chunk_len),
+            range: row_range,
+            format: self.texture_descriptor.format,
+        })
+    }
+
+    /// Common logic between [`Self::rect_bytes`] and [`Self::rect_bytes_mut`].
+    ///
+    /// Returns:
+    ///
+    /// 1. The absolute range in the data for the rect.
+    /// 2. The length of a row in the rect.
+    /// 3. The range within each row for the rect.
+    #[inline]
+    fn rect_bytes_inner(&self, rect: URect, z: u32) -> Option<(Range<usize>, usize, Range<usize>)> {
+        let width = self.texture_descriptor.size.width;
+        let start = self.pixel_data_offset(UVec3::new(0, rect.min.y, z))?;
+        let end = self.pixel_data_offset(UVec3::new(width, rect.max.y, z))?;
```

Comment:
This seems to return past the given rect in my tests when trying to pass a rect that should describe the whole Image. Notice the -1 here https://github.com/KirmesBude/bevy/commit/7542e4e8acac1262d0438d437126da7f19863e79#diff-b2f6268dff5264f58b7bcbe294d170114d810b2d6bdf8c86f9c371bbbc9422d9R118-R137

On that note, I think a convenience method on Image to translate the whole Image would be nice.

Reply:
Yeah, I was thinking of adding a method to convert the entire image to a `RectBytes`, but hadn't gotten there yet.